### PR TITLE
Arrangement 2: Enhance CCB merging with union set

### PR DESCRIPTION
**Pull request by sgiraudot**
## Summary of Changes

When merging 2 inner CCBs, all halfedges of one of the CCB are traversed again and updated, which can lead to an explosion of complexity, for example when computing the arrangement of a very large holes with many concavities.

This PR replaces this merging algorithm by the following:
- a `Union_find` structure of inner CCBs is used
- everytime inner CCBs are merged, we use `unify_sets()` on the union find
- the union find is used to access the correct inner CCB of each halfedge (to always fall back to the currently active CCB in the union find)
- after the sweep, the CCB are cleaned up using the union find to recover the correct result where each halfedge points to a valid inner CCB

In practice, this leads to accelerations up to 40x (1sec instead of 40sec) in the case similar to the one I described (large holes with many concavities). For more common data sets, I observed an acceleration of 30%.

As this enhancement comes at the cost of additional memory usage (one union find structure + 1 handle per inner CCB), I added the possibility to deactivate it with the macro `CGAL_ARRANGEMENT_DISABLE_UNION_SET_CCB`.

## Release Management

* Affected package(s): `Arrangement_2` + packages that depend on it (Boolean Op, Surface Sweep, etc.)

### CODE REVIEW

File: Arrangement_on_surface_2/include/CGAL/Arr_dcel_base.h [hunk 1]
```diff
@@ -34,6 +34,9 @@
 #include <CGAL/Arrangement_2/Arrangement_2_iterators.h>
 #include <CGAL/assertions.h>
 
+#ifndef CGAL_ARRANGEMENT_DISABLE_UNION_SET_CCB
```
Comment (lrineau) [inline Arrangement_on_surface_2/include/CGAL/Arr_dcel_base.h:37]: Do you really want to keep that macro `CGAL_ARRANGEMENT_DISABLE_UNION_SET_CCB` in CGAL?

If you do, you should add a test that actually defines it.

File: Arrangement_on_surface_2/include/CGAL/Arr_dcel_base.h [hunk 2]
```diff
@@ -762,10 +765,18 @@ class Arr_inner_ccb : public In_place_list_base<Arr_inner_ccb<V,H,F> >
   typedef Arr_face<V,H,F>                    Face;
   typedef typename Face::Inner_ccb_iterator  Inner_ccb_iterator;
 
+#ifndef CGAL_ARRANGEMENT_DISABLE_UNION_SET_CCB
+  typedef CGAL::Union_find<Self*>            Uf;
+  typedef typename Uf::handle                Uf_handle;
```
Comment (sloriot) [inline Arrangement_on_surface_2/include/CGAL/Arr_dcel_base.h:770]: you probably want to have a less generic name for the typedefs and variables

File: Arrangement_on_surface_2/include/CGAL/Surface_sweep_2/Arr_construction_ss_visitor.h [hunk 3]
```diff
@@ -142,6 +142,9 @@ class Arr_construction_ss_visitor :
   /* A notification issued before the sweep process starts. */
   inline void before_sweep();
 
+  /*! A notification issued after the sweep process ends. */
+  inline void after_sweep();
```
Comment (sloriot) [inline Arrangement_on_surface_2/include/CGAL/Surface_sweep_2/Arr_construction_ss_visitor.h:146]: why is this needed?

File: Arrangement_on_surface_2/include/CGAL/Arr_dcel_base.h [hunk 4]
```diff
@@ -930,13 +961,18 @@ class Arr_dcel_base {
   Inner_ccb_allocator in_ccb_alloc;         // An allocator for inner CCBs.
   Iso_vert_allocator  iso_vert_alloc;       // Allocator for isolated vertices.
 
+#ifndef CGAL_ARRANGEMENT_DISABLE_UNION_SET_CCB
+  Inner_ccb_uf        inner_ccb_uf;         // efficient CCB merging
```
Comment (sloriot) [inline Arrangement_on_surface_2/include/CGAL/Arr_dcel_base.h:965]: you probably want to clear the container when you're done with the construction of the arrangement.

File: Arrangement_on_surface_2/include/CGAL/Arr_dcel_base.h [hunk 5]
```diff
@@ -762,10 +765,18 @@ class Arr_inner_ccb : public In_place_list_base<Arr_inner_ccb<V,H,F> >
...
+  typedef CGAL::Union_find<Self*>            Uf;
+  typedef typename Uf::handle                Uf_handle;
+#endif
+
 private:
   Face* p_f;                  // The face the contains the CCB in its interior.
   Inner_ccb_iterator  iter;   // The inner CCB identifier.
   bool iter_is_not_singular;
+#ifndef CGAL_ARRANGEMENT_DISABLE_UNION_SET_CCB
+  Uf_handle ufh;
```
Comment (sloriot) [inline Arrangement_on_surface_2/include/CGAL/Arr_dcel_base.h:778]: I was trying to think about a way to avoid storing this extra handles but then I realized that it is only at the CCB level so it is probably useless trying to optimize that. If we want to optimise it, one possible way is to replace `p_f` by `std::union F_or_id { Face* p_f; std::size_t id };` and use the id to also store the face pointer in a vector while sweeping.

**Comment by lrineau** [comment]
You have to declare the new dependency to `Union_find`: https://travis-ci.com/github/CGAL/cgal/jobs/362900640#L7733

File: Arrangement_on_surface_2/test/Arrangement_on_surface_2/cgal_test.cmake [hunk 6]
```diff
@@ -659,6 +659,17 @@ function(test_construction_segments)
...
+#---------------------------------------------------------------------#
+# construction with segments, without Union Find
+#---------------------------------------------------------------------#
+function(test_construction_segments_no_union_find)
+  set(nt ${CGAL_GMPQ_NT})
+  set(kernel ${CARTESIAN_KERNEL})
+  set(geom_traits ${SEGMENT_GEOM_TRAITS})
+  set(flags "-DTEST_NT=${nt} -DTEST_KERNEL=${kernel} -DTEST_GEOM_TRAITS=${geom_traits} -DCGAL_ARRANGEMENT_DISABLE_UNION_SET_CCB")
+  compile_and_run_with_flags( test_construction segments_no_union_find "${flags}")
+endfunction()
```
Comment (lrineau) [inline Arrangement_on_surface_2/test/Arrangement_on_surface_2/cgal_test.cmake:671]: 1. You add a function, but that does not add any test if you do not all it.
2. The Arr_2 test suite is complex, and it has also a non-CTest version, in `cgal_test_with_cmake`.

File: Arrangement_on_surface_2/include/CGAL/Arr_dcel_base.h (hunk 2 above)
Comment (sgiraudot) [reply Arrangement_on_surface_2/include/CGAL/Arr_dcel_base.h:770]: Such as?

File: Arrangement_on_surface_2/include/CGAL/Surface_sweep_2/Arr_construction_ss_visitor.h (hunk 3 above)
Comment (sgiraudot) [reply Arrangement_on_surface_2/include/CGAL/Surface_sweep_2/Arr_construction_ss_visitor.h:146]: To call [`clean_inner_ccb()`](https://github.com/CGAL/cgal/pull/4851/files/1be446983801c00884d785ff9acc16fe5f913f75#diff-c697bc93f9992191b6ca431c9ee36d37R280).

File: Arrangement_on_surface_2/include/CGAL/Arr_dcel_base.h (hunk 4 above)
Comment (sgiraudot) [reply Arrangement_on_surface_2/include/CGAL/Arr_dcel_base.h:965]: Indeed, I'll change that.

**Comment by sgiraudot** [comment]
After talking it through with Sebastien, we decided to externalize as much as possible this Union find behavior so that it's only used by the surface sweep (to avoid creating invalid arrangements for users who manage it manually). I'll work on that, the code is likely to change quite a lot, so don't waste time reviewing right now :)

**Comment by sgiraudot** [comment]
Replaced by https://github.com/CGAL/cgal/pull/4875.
